typescript/channel.ts

```diff
@@ -50,16 +50,23 @@ export class Channel {
                     resolve(undefined);
                     return;
                 }
-                if (this.readBuf.length >= len) {
-                    let data = this.readBuf.slice(0, len);
-                    this.readBuf = this.readBuf.slice(len);
-                    resolve(data);
-                    if (this.readBuf.length == 0 && this.gotEOF) {
+                if (this.readBuf.length == 0) {
+                    if (this.gotEOF) {
                         this.readBuf = undefined;
+                        resolve(undefined);
+                        return;
                     }
+                    // TODO should these be served in-order? This will re-queue at
+                    // the end, but is that an issue?
```

Comment:
If there are multiple `read()` calls pending on the same channel, do you have any expectations that they'll be resolved in FIFO order? I guess for the most part this will happen on the first read attempt when there's no data in the buffer. Otherwise, this should get called again when there's actually data to read. But should there be an explicit guarantees about the order, or just say that "concurrent" reads are undefined for the order they'll be resolved.

Reply:
yes i think FIFO is expected, though not intended for multiple readers